cryptofeed/exchanges/mixins/bitfinex_rest.py

```diff
@@ -12,31 +12,43 @@
 
 from yapic import json
 
-from cryptofeed.defines import BID, ASK, BUY, L2_BOOK, L3_BOOK, SELL, TICKER, TRADES
+from cryptofeed.defines import (
+    BID, ASK, BUY, L2_BOOK, L3_BOOK, SELL, TICKER, TRADES, MARKET, LIMIT,
+    MARGIN_LIMIT, MARGIN_MARKET, CANCEL_ORDER, PLACE_ORDER,
+)
 from cryptofeed.exchange import RestExchange
 from cryptofeed.util.time import timedelta_str_to_sec
 from cryptofeed.types import OrderBook, Candle
 
 
 class BitfinexRestMixin(RestExchange):
     api = "https://api-pub.bitfinex.com/v2/"
+    auth_api = 'https://api.bitfinex.com'
     rest_channels = (
-        TRADES, TICKER, L2_BOOK, L3_BOOK
+        TRADES, TICKER, L2_BOOK, L3_BOOK, CANCEL_ORDER, PLACE_ORDER
     )
+    order_options = {
+        LIMIT: 'EXCHANGE LIMIT',
+        MARKET: 'EXCHANGE MARKET',
+        MARGIN_LIMIT: 'LIMIT',
+        MARGIN_MARKET: 'MARKET'
+    }
     candle_mappings = {'1m': '1m', '5m': '5m', '15m': '15m', '30m': '30m', '1h': '1h', '3h': '3h', '6h': '6h', '12h': '12h', '1d': '1D', '1w': '7D', '2w': '14D', '1M': '1M'}
 
     def _nonce(self):
-        return str(int(round(time.time() * 1000)))
+        return str(int(round(time.time() * 1000000)))
 
-    def _generate_signature(self, url: str, body=json.dumps({})):
+    def _generate_signature(self, url: str, body=None):
+        if not body:
+            body = json.dumps({})
         nonce = self._nonce()
         signature = "/api/" + url + nonce + body
-        h = hmac.new(self.config.key_secret.encode('utf8'), signature.encode('utf8'), hashlib.sha384)
+        key_secret = self.config['BITFINEX']['key_secret'].encode('utf8')
+        h = hmac.new(key_secret, signature.encode('utf8'), hashlib.sha384)
         signature = h.hexdigest()
-
         return {
             "bfx-nonce": nonce,
-            "bfx-apikey": self.config.key_id,
+            "bfx-apikey": self.config['BITFINEX']['key_id'],
```

Comment:
I believe this should just be self.key_id. They key information is loaded into the exchange object upon instantiation.